EREGCSC-314 Create SubpartReaderView

cgodwin1 opened the pull request.

Commit:
Added path converters and created SubpartReaderView

* Path converters are useful for more explicitly defining our paths
* SubpartReaderView may or may not be the correct answer to EREGCSC-314

Commit:
Revised converters and fixed subpart view bug

* Converters now inherit from PathConverter for simplicity.
* Combined Part- and SectionConverter to single NumericConverter.
* With 'subpart' variable in URL SubpartReaderView would not display properly.

Commit:
Renamed reader_view to section_reader_view for clarity

Commit:
fixup

Commit:
fixup (hopefully)

whytheplatypus commented on regulations/views/mixins.py:
```diff
@@ -14,6 +14,8 @@ def build_citation(context):
             citation.append(context["part"])
         if 'section' in context:
             citation.append(context["section"])
+        if 'subpart' in context:
+            citation.append(context["subpart"])
```
At the moment a `citation` doesn't contain both a section and subpart. Add some logic here to make model that.

cgodwin1 replied:
Resolved in d865569

Commit:
Rewrote TableOfContentsMixin for desired behavior

TOC mixin now takes into consideration the requested view,
and is able to generate a table of contents that links to subpart,
part, or section reader views.

Commit:
Cleanup for TableOfContentsMixin

Commit:
Added hash for part-section in table of contents

Allows jumping to sections without leaving the page in
subpart and part reader views.

Commit:
Added states for part, subpart, and section view links

Commit:
Added functional links for SubpartReaderView

Commit:
Added active class to currently active view link

whytheplatypus commented on regulations/views/mixins.py:
```diff
@@ -53,15 +56,39 @@ def build_sidebar_context(self, sidebar_class, label_id, version):
+        part = citation[0]
+        section = citation[1]
+
+        if view_name == "subpart_reader_view" and subpart is not None:
+            view = view_name
+            args = (part, subpart, version)
+        elif view_name == "part_reader_view":
+            view = view_name
+            args = (part, version)
+        else:
+            view = self.default_view
+            args = (part, section, version)
```
This conditional logic can move into the individual view classes now via an override of `get_url`

cgodwin1 replied:
Implemented in 5f17599

whytheplatypus commented on regulations/views/reader.py:
```diff
@@ -24,19 +25,19 @@ def get_context_data(self, **kwargs):
         reg_part = context["part"]
         reg_citation = context["citation"]
-        toc = self.get_toc(reg_part, reg_version)
+        toc = self.get_toc(context['part'], context['version'])
```
Can this continue to use the `reg_part` and `reg_citation` variables?

cgodwin1 replied:
Resolved in a2e2a69

whytheplatypus commented on regulations/views/reader.py:
```diff
@@ -59,3 +60,61 @@ def get_neighboring_sections(self, label, version):
+
+        part = context['part']
+        version = context['version']
+        first_section = utils.first_section(part, version)
+        first_subpart = utils.first_subpart(part, version)
+
+        c = {
+            'subpart_view_link': reverse('subpart_reader_view', args=(part, first_subpart, version)),
+            'section_view_link': reverse('section_reader_view', args=(part, first_section, version)),
+        }
```
move this into a function with a common name across the classes.

cgodwin1 replied:
Implemented in c8537d2

whytheplatypus reviewed (changes requested):
clicked the wrong button :facepalm:

Commit:
Moved view switch link generation to an abstract method

Commit:
Cleanup

Commit:
Moved TOC URL building override to reader classes

Commit:
Added error cases to build_citation(...)

Commit:
Added smooth scrolling behavior

Commit:
fixup

Commit:
Attempt fix on tests

Assumption: tests failing because version variable in URL
now constrained with regex to be X-X where X is alphanumeric.

Commit:
Fixed tests!

* print_part.html was reversing section_reader_view with only a part argument.
* views_redirect_tests.py was using one-word versions for reversing.

Commit:
Linting fixup

whytheplatypus commented on regulations/views/mixins.py:
```diff
@@ -2,18 +2,23 @@
         citation = []
         if 'part' in context:
             citation.append(context["part"])
+        else:
+            return []
```
I messed this up. The function shouldn't return two different types, so this should return an empty string or the following conditionals should be nested inside this if e.g.

```python
        if 'part' in context:
            citation.append(context["part"])
            if 'section' in context:
                citation.append(context["section"])
            elif 'subpart' in context:
                citation.append(context["subpart"])
        return "-".join(citation)
```

cgodwin1 replied:
Resolved in b50e190

Commit:
Fixed build_citation to return string not array
